lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +796,585 @@ def __str__(self):
+        return 'refs/heads/{0}'.format(self._ref)
+
+    def fetch_spec(self):
+        # type: () -> str
+        """Return an argument to update a local ref from the remote, if applicable."""
+        if self.ref_type == 'Commit':
+            # Commits cannot be updated and do not require any special syntax to fetch.
+            return self.refspec()
+        # Otherwise, we use the syntax to update the local ref from a remote ref of
+        # the same name, fetching updates if available/applicable.
+        assert self.ref_type in ['Tag', 'Branch'], self
+        maybe_mutating_prefix = '+' if self.is_mutable() else ''
+        return '{0}{1}:{1}'.format(maybe_mutating_prefix, self.refspec())
+
+    def is_mutable(self):
+        # type: () -> bool
+        """Whether this type of git ref should check for updates to that ref."""
+        if self.ref_type == 'Branch':
+            return True
+        assert self.ref_type in ['Tag', 'Commit'], self
+        return False
+
+    def _repo_info(self):
+        # type: () -> str
+        if self.ref_type == 'Commit':
+            return 'at commit {0}'.format(self._ref)
+        if self.ref_type == 'Tag':
+            return 'at tag {0}'.format(self._ref)
+        assert self.ref_type == 'Branch', self
+        return 'on branch {0}'.format(self._ref)
+
+    def repo_info_for_reference(self):
+        # type: () -> str
+        if self.should_validate_matches_hash:
+            return '{} (at commit {})'.format(
+                self._repo_info(), self.should_validate_matches_hash)
+        return self._repo_info()
+
+
+class GitCommit(GitRef):
+    """A git commit."""
+
+    ref_type = 'Commit'        # type: ClassVar[str]
+
+    # Matches any 7-40 character hexadecimal string.
+    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
+
+    @property
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        """There is no need to verify that a git commit hash matches itself."""
+        return None
+
+    def __init__(self, ref):
+        # type: (str) -> None
+        if not isinstance(ref, six.string_types):
+            # The exception message from `re.match()` does not print the value when
+            # a non-string argument is provided, so we special-case that check here.
+            raise InvalidGitRef('git reference was not a string: {0}'.format(ref))
+        if not self._commit_rx.match(ref):
+            raise InvalidGitRef(dedent("""\
+            Spack requires that references to individual git commits be specified via
+            a 7-40 character hexadecimal string, but received {0!r} instead.
+
+            A valid commit hash case-insensitively matches the regular expression '{1}'.
+            7 hex characters is the size printed out by `git log --format='%h'`, while
+            40 hex characters is the size printed out by `git log --format='%H'`.
+            """.format(ref, self._commit_rx.pattern)))
```

Comment:
We should check this isn't already provided (the regex) somewhere else in spack.

Reply:
I tried to ensure we centralized the regex patterns in this PR! I'm pretty sure this is provided just in two places, and they're slightly different patterns:
```bash
# cosmicexplorer@terrestrial-gamma-ray-flash: ~/tools/s1 18:05:59 
; rg -i '0\-9a\-f'
lib/spack/spack/fetch_strategy.py
931:    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
1197:    _tag_mapping_rx = re.compile(r'^([0-9a-f]{40}) refs/tags/(.*)$')

share/spack/qa/flake8_formatter.py
51:            r'([\'"])[0-9a-fA-F]{32,}\1',  # long hex checksums

lib/spack/external/ruamel/yaml/resolver.py
213:    |[-+]?0x[0-9a-fA-F_]+
281:        |[-+]?0x[0-9a-fA-F_]+)$''', re.X),
288:        |[-+]?0x[0-9a-fA-F_]+

asdf/spack-src/build-aux/gitlog-to-changelog
169:          $line =~ /^([0-9a-fA-F]{40})$/
243:      $sha =~ /^[0-9a-fA-F]{40}$/

var/spack/repos/builtin/packages/binutils/update_symbol-2.26.patch
70:+ +[0-9]+: +[0-9a-f]+ +[0-9a-f]+ +FUNC +GLOBAL +DEFAULT +[0-9]+ +foo@VERS.1
72:+ +[0-9]+: +[0-9a-f]+ +[0-9a-f]+ +FUNC +GLOBAL +DEFAULT +[0-9]+ +foo@@VERS.2
```

where `_tag_mapping_rx` specifically searches for 40-character hexadecimal strings, and *not* `7-40` characters like `_commit_rx`. We could merge it with `_commit_rx`, but I personally appreciate having all that matching logic stored in the `_tag_mapping_rx` pattern.

Were you thinking of anywhere else that should be doing this logic differently?